SAP/cloud-sdk
# chore: add OData V4 e2e tests

## Context
Add a new e2e test stage for verifying OData V4 implementation. 
The reason is simple, the existing integration test strategy where we mocked the requests/response, is not enough.

### which OData v4 service are we using?
We build a service ourselves by using cap with an in-memory db, because the public service is not stable enough.

### what's new in the GitHub action?
- e2e test stage
  - build the app (generate db file for sqlite)
  - start the app (wrapped in `pm2` to make it as daemon mode)
  - run the jest tests

### how do we test?
- start the server locally, open the metadata page, download it and put it into the `test-resources` [1]
- generate the vdms by using the same command as you used for test entities. [2]
- write your test in the `e2e-test` just like the integration tests.

The [1] and [2] are not automated, which is the same as the quarterly release and the test entities. It should be stable.

### any other highlights?
- the whole e2e tests takes < 100s in the GitHub Actions, which can be improved further when using cache.
- the poc projects also tried to used cypress GitHub action, which can be used for frontend e2e tests later, also only needed < 100s.
- the `@sap/cds` lib is fetched from the npmjs.org

### future planning
- if the e2e tests takes longer or becomes flaky , we can consider disable it for internal developers, but make it as prerequisites of the release pipeline.

## Definition of Done

Please consider all items and remove only if not applicable.

- [x] Tests created/adjusted for your changes.
  * Provide sufficient context so that each entry can be understood on its own.
  * Be specific about names of functions, classes, modules, etc.
  * Describe when or where this is relevant
  * Use indicative and present tense. For example, write "Provide function `name` that does X in order to Y" over "Now X can be done by calling a new function".
- [x] PR title adheres to [conventional commit guidelines](https://www.conventionalcommits.org) (please note that only `fix:` and `feat:` will end up in the release notes)
- [ ] If applicable: Properly documented (JSDoc of public API)
- [ ] If applicable: Check if `node run doc` still works.
